### PR TITLE
Refactor OCPAPIServer to conform to the internal service manager

### PR DESCRIPTION
**Pull request by husky-parul**
Signed-off-by: Parul <parsingh@redhat.com>

Refactor OCPAPIServer to conform to the internal service manager

### CODE REVIEW

File: pkg/controllers/ocp-controllers.go [hunk 1]
```diff
@@ -38,80 +37,98 @@ import (
...
+	defer close(stopped)
+
 	go func() {
-		logrus.Fatalf("ocp apiserver exited: %v", command.Execute())
+		err := s.prepareOCPComponents()
+		if err != nil {
+			logrus.Warningf("Failed to prepare OCP components")
+		}
+		// run healthcheck
+		healthcheckStatus := util.RetryTCPConnection("127.0.0.1", "8444")
```
Comment (rootfs) [inline pkg/controllers/ocp-controllers.go:97]: this is not needed, `waitForOCPAPIServer` is actually the health check

File: pkg/controllers/ocp-controllers.go [hunk 2]
```diff
@@ -38,80 +37,98 @@ import (
...
+
+func (s *OCPAPIServer) configure(cfg *config.MicroshiftConfig) error {
+	var configFilePath = cfg.DataDir + "/resources/openshift-apiserver/config/config.yaml"
+
+	if err := config.OpenShiftAPIServerConfig(cfg); err != nil {
+		logrus.Infof("Failed to create a new ocp-apiserver configuration: %v", err)
+		return err
+	}
 	args := []string{
-		"start",
```
Comment (rootfs) [inline pkg/controllers/ocp-controllers.go:60]: what happen to these args?

Comment (husky-parul) [reply pkg/controllers/ocp-controllers.go:60]: They are read from the config file when `RunAPIServer` is called  
[here](https://github.com/redhat-et/microshift/blob/e9c795bf061a2bc7dce47adc03e1965c2281bfe0/pkg/controllers/ocp-controllers.go#L108)

File: pkg/controllers/ocp-controllers.go [hunk 3]
```diff
@@ -38,80 +37,95 @@ import (
...
+		logrus.Fatalf("Failed to start ocp-apiserver %v", err)
+	}
+	return ctx.Err()
+}
+func (s *OCPAPIServer) prepareOCPComponents() error {
 
 	// ocp api service registration
-	if err := createAPIHeadlessSvc(cfg); err != nil {
-		logrus.Warningf("failed to apply headless svc %v", err)
-		return err
```
Comment (rootfs) [inline pkg/controllers/ocp-controllers.go:88]: not drop errors

File: pkg/controllers/ocp-controllers.go [hunk 4]
```diff
@@ -38,80 +37,95 @@ import (
...
+	if err := createAPIHeadlessSvc(s.cfg); err != nil {
+		logrus.Warningf("Failed to apply headless svc %v", err)
 	}
-	if err := createAPIRegistration(cfg); err != nil {
-		logrus.Warningf("failed to register api %v", err)
-		return err
```
Comment (rootfs) [inline pkg/controllers/ocp-controllers.go:92]: same here

File: pkg/controllers/ocp-controllers.go [hunk 5]
```diff
@@ -38,80 +37,95 @@ import (
...
+	if err := createAPIRegistration(s.cfg); err != nil {
+		logrus.Warningf("Failed to register api %v", err)
 	}
-
 	// probe ocp api services
-	restConfig, err := clientcmd.BuildConfigFromFlags("", cfg.DataDir+"/resources/kubeadmin/kubeconfig")
+	restConfig, err := clientcmd.BuildConfigFromFlags("", s.cfg.DataDir+"/resources/kubeadmin/kubeconfig")
 	if err != nil {
-		return err
```
Comment (rootfs) [inline pkg/controllers/ocp-controllers.go:98]: not drop errors

**Comment by rootfs** [comment]
can you run the following manual tests on your end to see if the resources and services are ready? 
```console
# kubectl api-resources |grep openshift
rolebindingrestrictions                        authorization.openshift.io/v1            true         RoleBindingRestriction
builds                                         config.openshift.io/v1                   false        Build
images                                         config.openshift.io/v1                   false        Image
configs                                        imageregistry.operator.openshift.io/v1   false        Config
imagecontentsourcepolicies                     operator.openshift.io/v1alpha1           false        ImageContentSourcePolicy
securitycontextconstraints                     security.openshift.io/v1                 false        SecurityContextConstraints
# # kubectl get apiservices |grep openshift
v1.apps.openshift.io                     default/openshift-apiserver   True        26s
v1.authorization.openshift.io            Local                         True        51s
v1.build.openshift.io                    default/openshift-apiserver   True        26s
v1.config.openshift.io                   Local                         True        41s
v1.image.openshift.io                    default/openshift-apiserver   True        26s
v1.imageregistry.operator.openshift.io   Local                         True        56s
v1.project.openshift.io                  default/openshift-apiserver   True        26s
v1.quota.openshift.io                    Local                         True        26s
v1.route.openshift.io                    default/openshift-apiserver   True        26s
v1.security.openshift.io                 Local                         True        61s
v1.template.openshift.io                 default/openshift-apiserver   True        25s
v1alpha1.operator.openshift.io           Local                         True        46s
```

**Comment by husky-parul** [comment]
> ```
> kubectl get apiservices |grep openshift
> ```

```
sudo kubectl get apiservices --kubeconfig=/var/lib/microshift/resources/kubeadmin/kubeconfig | grep openshift
v1.apps.openshift.io                     default/openshift-apiserver   True        3m42s
v1.authorization.openshift.io            default/openshift-apiserver   True        3m42s
v1.build.openshift.io                    default/openshift-apiserver   True        3m42s
v1.config.openshift.io                   Local                         True        3m12s
v1.image.openshift.io                    default/openshift-apiserver   True        3m42s
v1.imageregistry.operator.openshift.io   Local                         True        3m27s
v1.project.openshift.io                  default/openshift-apiserver   True        3m42s
v1.quota.openshift.io                    default/openshift-apiserver   True        3m42s
v1.route.openshift.io                    default/openshift-apiserver   True        3m42s
v1.security.openshift.io                 default/openshift-apiserver   True        3m41s
v1.template.openshift.io                 default/openshift-apiserver   True        3m41s
v1alpha1.operator.openshift.io           Local                         True        3m17s
```

**Comment by husky-parul** [comment]
I see that `project apiservices` are created
```
oc get apiservices | grep project
v1.project.openshift.io                  default/openshift-apiserver         True        97m
```
But on creating and accessing the project is an issue 
```
oc get projects 
Error from server (Forbidden): project.project.openshift.io is forbidden: unable to list projects without a user on the context
``
@oglok looks like it has to do with `sccs` not getting created. Will report back.

File: pkg/controllers/openshift-apiserver.go [hunk 6]
```diff
@@ -32,48 +33,91 @@ import (
...
+func (s *OCPAPIServer) Run(ctx context.Context, ready chan<- struct{}, stopped chan<- struct{}) error {
+	defer close(stopped)
 
-	command.SetArgs(args)
-	logrus.Infof("starting openshift-apiserver, args: %v", args)
-	go func() {
-		logrus.Fatalf("ocp apiserver exited: %v", command.Execute())
-	}()
+	// go func() {
+	// 	err := s.prepareOCPComponents(s.cfg)
```
Comment (rootfs) [inline pkg/controllers/openshift-apiserver.go:103]: you can refactor `prepareOCPComponents` and move `waitForOCPAPIServer` to here as health checker

File: pkg/controllers/openshift-apiserver.go [hunk 7]
```diff
@@ -32,48 +33,91 @@ import (
...
+	// 	if err != nil {
+	// 		logrus.Warningf("Failed to start ocp-apiserver %v", err)
+	// 	}
+	// 	logrus.Infof("%s is ready", s.Name())
+
+	// 	close(ready)
```
Comment (rootfs) [inline pkg/controllers/openshift-apiserver.go:109]: need to signal readiness

**Comment by husky-parul** [comment]
@rootfs @oglok 
Please take a final look. 
[Commit](https://github.com/redhat-et/microshift/pull/324/commits/8e9e6fafd4b3551ea6152d7342d56965f9e20f24) fixed the problem. 

I've done the basic testing for openshift and it looks good to me. 

```
oc new-project test 
Now using project "test" on server "https://127.0.0.1:6443".

oc get projects 
NAME                            DISPLAY NAME   STATUS
default                                        Active
demo                                           Active
kube-node-lease                                Active
kube-public                                    Active
kube-system                                    Active
kubevirt-hostpath-provisioner                  Active
openshift                                      Active
openshift-controller-manager                   Active
openshift-dns                                  Active
openshift-infra                                Active
openshift-ingress                              Active
openshift-node                                 Active
openshift-service-ca                           Active
test                                           Active

oc get apiservices   | grep openshift 
v1.apps.openshift.io                     default/openshift-apiserver         True        29m
v1.authorization.openshift.io            Local                               True        3m47s
v1.build.openshift.io                    default/openshift-apiserver         True        29m
v1.config.openshift.io                   Local                               True        3m47s
v1.image.openshift.io                    default/openshift-apiserver         True        29m
v1.imageregistry.operator.openshift.io   Local                               True        3m47s
v1.project.openshift.io                  default/openshift-apiserver         True        29m
v1.quota.openshift.io                    Local                               True        29m
v1.route.openshift.io                    default/openshift-apiserver         True        29m
v1.security.openshift.io                 Local                               True        3m47s
v1.template.openshift.io                 default/openshift-apiserver         True        29m
v1.user.openshift.io                     default/openshift-oauth-apiserver   True        29m
v1alpha1.operator.openshift.io           Local                               True        3m47s

oc  api-resources   | grep openshift 
deploymentconfigs                 dc           apps.openshift.io/v1                     true         DeploymentConfig
rolebindingrestrictions                        authorization.openshift.io/v1            true         RoleBindingRestriction
buildconfigs                      bc           build.openshift.io/v1                    true         BuildConfig
builds                                         build.openshift.io/v1                    true         Build
builds                                         config.openshift.io/v1                   false        Build
images                                         config.openshift.io/v1                   false        Image
proxies                                        config.openshift.io/v1                   false        Proxy
images                                         image.openshift.io/v1                    false        Image
imagesignatures                                image.openshift.io/v1                    false        ImageSignature
imagestreamimages                 isimage      image.openshift.io/v1                    true         ImageStreamImage
imagestreamimports                             image.openshift.io/v1                    true         ImageStreamImport
imagestreammappings                            image.openshift.io/v1                    true         ImageStreamMapping
imagestreams                      is           image.openshift.io/v1                    true         ImageStream
imagestreamtags                   istag        image.openshift.io/v1                    true         ImageStreamTag
imagetags                         itag         image.openshift.io/v1                    true         ImageTag
configs                                        imageregistry.operator.openshift.io/v1   false        Config
imagecontentsourcepolicies                     operator.openshift.io/v1alpha1           false        ImageContentSourcePolicy
projectrequests                                project.openshift.io/v1                  false        ProjectRequest
projects                                       project.openshift.io/v1                  false        Project
clusterresourcequotas                          quota.openshift.io/v1                    false        ClusterResourceQuota
routes                                         route.openshift.io/v1                    true         Route
securitycontextconstraints                     security.openshift.io/v1                 false        SecurityContextConstraints
brokertemplateinstances                        template.openshift.io/v1                 false        BrokerTemplateInstance
processedtemplates                             template.openshift.io/v1                 true         Template
templateinstances                              template.openshift.io/v1                 true         TemplateInstance
templates                                      template.openshift.io/v1                 true         Template
groups                                         user.openshift.io/v1                     false        Group
identities                                     user.openshift.io/v1                     false        Identity
useridentitymappings                           user.openshift.io/v1                     false        UserIdentityMapping
users                                          user.openshift.io/v1                     false        User
```

File: pkg/controllers/infra-services-controller.go [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+func (s *InfrastructureServicesManager) Name() string { return "infrastructure-services-manager" }
+func (s *InfrastructureServicesManager) Dependencies() []string {
+	return []string{"kube-apiserver", "openshift-prepjob-manager", "oauth-apiserver", "openshift-controller-manager"}
+}
+
+func (s *InfrastructureServicesManager) Run(ctx context.Context, ready chan<- struct{}, stopped chan<- struct{}) error {
+	defer close(stopped)
+
+	go func() error {
```
Comment (rootfs) [inline pkg/controllers/infra-services-controller.go:45]: this goroutine is not needed, can be removed

File: pkg/controllers/openshift-api-components-controller.go [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+	return "openshift-api-components-manager"
+}
+func (s *OpenShiftAPIComponentsControllerManager) Dependencies() []string {
+	return []string{"kube-apiserver", "openshift-prepjob-manager", "oauth-apiserver"}
+}
+
+func (s *OpenShiftAPIComponentsControllerManager) Run(ctx context.Context, ready chan<- struct{}, stopped chan<- struct{}) error {
+	defer close(stopped)
+
+	go func() {
```
Comment (rootfs) [inline pkg/controllers/openshift-api-components-controller.go:46]: this goroutine is not needed, just need to close ready channel

File: pkg/controllers/infra-services-controller.go [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+	return s
+}
+
+func (s *InfrastructureServicesManager) Name() string { return "infrastructure-services-manager" }
+func (s *InfrastructureServicesManager) Dependencies() []string {
+	return []string{"kube-apiserver", "openshift-prepjob-manager", "oauth-apiserver", "openshift-controller-manager"}
+}
+
+func (s *InfrastructureServicesManager) Run(ctx context.Context, ready chan<- struct{}, stopped chan<- struct{}) error {
+	defer close(stopped)
```
Comment (rootfs) [inline pkg/controllers/infra-services-controller.go:43]: unlike other daemonized services, this function will return, should this channel be closed? @fzdarsky

File: pkg/controllers/openshift-api-components-controller.go [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+
+func (s *OpenShiftAPIComponentsControllerManager) Name() string {
+	return "openshift-api-components-manager"
+}
+func (s *OpenShiftAPIComponentsControllerManager) Dependencies() []string {
+	return []string{"kube-apiserver", "openshift-prepjob-manager", "oauth-apiserver"}
+}
+
+func (s *OpenShiftAPIComponentsControllerManager) Run(ctx context.Context, ready chan<- struct{}, stopped chan<- struct{}) error {
+	defer close(stopped)
```
Comment (rootfs) [inline pkg/controllers/openshift-api-components-controller.go:44]: unlike other daemonized services, this function will return, should stopped channel be closed? @fzdarsky

File: pkg/controllers/openshift-prepjob-manager.go [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+	return s
+}
+
+func (s *OCPPrepJobManager) Name() string { return "openshift-prepjob-manager" }
+func (s *OCPPrepJobManager) Dependencies() []string {
+	return []string{"kube-apiserver", "oauth-apiserver"}
+}
+
+func (s *OCPPrepJobManager) Run(ctx context.Context, ready chan<- struct{}, stopped chan<- struct{}) error {
+	defer close(stopped)
```
Comment (rootfs) [inline pkg/controllers/openshift-prepjob-manager.go:42]: unlike other daemonized services, this function will return, should this channel be closed? @fzdarsky

File: pkg/controllers/openshift-prepjob-manager.go [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+
+	go func() {
+		// To-DO add readiness check
+	}()
```
Comment (rootfs) [inline pkg/controllers/openshift-prepjob-manager.go:46]: goroutine is not needed, just close ready channel

File: pkg/controllers/openshift-prepjob-manager.go [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+
+func NewOpenShiftPrepJob(cfg *config.MicroshiftConfig) *OCPPrepJobManager {
+	s := &OCPPrepJobManager{}
+	s.cfg = cfg
+	return s
+}
+
+func (s *OCPPrepJobManager) Name() string { return "openshift-prepjob-manager" }
+func (s *OCPPrepJobManager) Dependencies() []string {
+	return []string{"kube-apiserver", "oauth-apiserver"}
```
Comment (rootfs) [inline pkg/controllers/openshift-prepjob-manager.go:38]: `PrepareOCP` should happen before `oauth-apiserver`

File: pkg/controllers/openshift-oauth-server.go [hunk 15]
```diff
@@ -41,8 +41,10 @@ func NewOpenShiftOAuth(cfg *config.MicroshiftConfig) *OpenShiftOAuth {
 	return s
 }
 
-func (s *OpenShiftOAuth) Name() string           { return "oauth-apiserver" }
-func (s *OpenShiftOAuth) Dependencies() []string { return []string{"kube-apiserver"} }
+func (s *OpenShiftOAuth) Name() string { return "oauth-apiserver" }
+func (s *OpenShiftOAuth) Dependencies() []string {
+	return []string{"kube-apiserver", "openshift-controller-manager"}
```
Comment (rootfs) [inline pkg/controllers/openshift-oauth-server.go:46]: also depends on ocp-apiserver

**Comment by mangelajo** [comment]
this looks almost ready @husky-parul let's get @rootfs comments addressed and then merged :rocket:

**Comment by husky-parul** [comment]
@rootfs all comments addressed. Ready to merge

**Pull request merged**